plugins/trip.hpp

```diff
@@ -0,0 +1,332 @@
+        // in_range: index where a new component starts
+        //
+        // example: NodeID 0, 1, 2, 4, 5 are in component 0
+        //          NodeID 3, 6, 7, 8    are in component 1
+        //          => in_component = [0, 1, 2, 4, 5, 3, 6, 7, 8]
+        //          => in_range = [0, 5]
+        SCC_Component(std::vector<NodeID> in_component,
+                      std::vector<size_t> in_range)
+                      : component(in_component),
+                        range(in_range) {
+            range.push_back(in_component.size());
+            BOOST_ASSERT_MSG(in_component.size() >= in_range.size(),
+                             "scc component and its ranges do not match");
+            BOOST_ASSERT_MSG(*std::max_element(in_range.begin(), in_range.end()) < in_component.size(),
+                             "scc component ranges are out of bound");
+            BOOST_ASSERT_MSG(*std::min_element(in_range.begin(), in_range.end()) >= 0,
+                             "invalid scc component range");
+            BOOST_ASSERT_MSG([&in_range](){
+                                 for (std::size_t r = 0; r < in_range.size() - 1; ++r) {
+                                     if (in_range[r] > in_range[r+1]) {
+                                         return false;
+                                     }
+                                 }
+                                 return true;
+                             }(),
+                             "invalid component ranges");
+          };
+
+        // constructor to use when whole graph is one single scc
+        SCC_Component(std::vector<NodeID> in_component)
+                      : component(in_component), range({0, in_component.size()}) {
+                      };
+
+        std::size_t GetNumberOfComponents() const{
+            return range.size() - 1;
+        }
+
+        const std::vector<NodeID> component;
+        // component range = in_range + [component.size()]
+        std::vector<size_t> range;
+    };
+
+    // takes the number of locations and its distance matrix,
+    // identifies and splits the graph in its strongly connected components (scc)
+    // and returns an SCC_Component
+    SCC_Component SplitUnaccessibleLocations(const std::size_t number_of_locations,
+                                             const DistTableWrapper<EdgeWeight> & result_table) {
+
+        // Run TarjanSCC
+        auto wrapper = std::make_shared<MatrixGraphWrapper<EdgeWeight>>(result_table.GetTable(), number_of_locations);
+        auto scc = TarjanSCC<MatrixGraphWrapper<EdgeWeight>>(wrapper);
+        scc.run();
+
+        std::vector<size_t> range_insertion;
+        std::vector<size_t> range;
+        range_insertion.reserve(scc.get_number_of_components());
+        range.reserve(scc.get_number_of_components());
+
+        std::vector<NodeID> components(number_of_locations, 0);
+
+        auto prefix = 0;
+        for (size_t j = 0; j < scc.get_number_of_components(); ++j){
```

Comment:
You have multiple `scc.get_number_of_components()` calls here. I would help the compiler and do common subexpression elimination here, that is store the result as `const` somewhere, instead of repeated function calls.